### PR TITLE
fix(kimi-k25): preserve thinking content in offline data formatting

### PR DESCRIPTION
**Pull request by qywu**
## Summary
`KimiK25Parser.format()` corrupts assistant turns from **thinking models** (Kimi K2.x) when their text is re-tokenized for **offline** training data. The opening `<think>` is emitted by the chat template as part of the *generation prompt*, so a captured response is `{reasoning}</think>{answer}` (closing tag, no opener) — or, with a reasoning parser, the reasoning is in a separate field and `content` is the bare answer. In both cases `format()` produced a malformed turn and silently corrupted the supervised sequence / loss mask.

This is latent on `main` because the **online/decode** path captures generated tokens verbatim (no re-tokenization); it only surfaces in **offline** EAGLE3 data prep.

## Changes
- **`_recover_missing_think_open`** — restore a dropped opening `<think>` when `content` has a `</think>` but no opener (raw, no reasoning parser).
- **`_merge_reasoning_field`** — rebuild `<think>{reasoning}</think>{answer}` from a separate `reasoning_content`/`thinking` field (reasoning-parser output, e.g. `--reasoning-parser kimi_k2`); fields kept in sync with `has_thinking_content`.
- Apply recovery on **non-last** turns *before* `_strip_thinking`, so dropped-opener historical turns strip cleanly (multi-turn).
- **`has_unbalanced_thinking_tags`** + an aggregated **load-time warning** in `load_conversation_dataset` (scoped to the offline path, so the decode generation-prompt `<think>` doesn't false-positive).

The two recovery paths compose and are mutually exclusive (inline vs separated). Plain answers still get `<think></think>`; well-formed content is unchanged.

## Behavior: before → after (last assistant turn unless noted)

| # | data source | reasoning carried in | upstream (before) | this PR (after) |
|---|---|---|---|---|
| 1 | no reasoning parser | inline `content` = `{reasoning}</think>{answer}` | ❌ `<think></think>{reasoning}</think>{answer}` (dangling `</think>`, reasoning outside think block) | ✅ `<think>{reasoning}</think>{answer}` |
| 2 | reasoning parser (`--reasoning-parser kimi_k2`) | separate `reasoning_content`/`thinking` field | ❌ `<think></think>{answer}` (reasoning silently dropped) | ✅ `<think>{reasoning}</think>{answer}` |
| 3 | inline well-formed | `content` = `<think>…</think>…` | ✅ unchanged | ✅ unchanged |
| 4 | plain answer (no reasoning) | `content` = `{answer}` | ✅ `<think></think>{answer}` | ✅ `<think></think>{answer}` |
| 5 | earlier (non-last) turns — well-formed / field / **dropped-opener** | either | ❌ dropped-opener stays malformed | ✅ thinking stripped (all cases) |

## Tests
`TestKimiK25ThinkRecovery`, `TestKimiK25ReasoningField`, `TestKimiK25MultiTurnDroppedOpener`, `TestThinkBalanceCheck` — 44 parser tests pass.

> Found while building an offline EAGLE3 overfit for Kimi-K2.6-NVFP4.

### CODE REVIEW

File: torchspec/data/parse.py [hunk 1]
```diff
@@ -404,6 +419,54 @@ def _format_content(self, content: str, role: str) -> str:
...
+        the closing ``</think>`` but not the opening one. Without recovery, such
+        content (which does not start with ``<think>``) gets an empty
+        ``<think></think>`` prepended below, producing a malformed turn:
+        ``<think></think>{reasoning}</think>{answer}`` with a dangling close tag
+        and the reasoning outside any think block. If a ``</think>`` is present
+        without a matching opener, restore the opener so the turn is a proper
+        ``<think>{reasoning}</think>{answer}`` block matching the model's output.
+        """
+        if "</think>" in content and not content.lstrip().startswith("<think>"):
+            return "<think>" + content
```
[the remark this answers is not in the document]
Comment (qywu) [reply torchspec/data/parse.py]: Good catch — confirmed and fixed in fdaa2fd.

`has_thinking_content` (used by `last_turn_loss_only='auto'`) only matched a literal opening `<think>` or a separate reasoning field, so the recovered `{reasoning}</think>{answer}` shape was treated as non-thinking and got loss on every assistant turn instead of only the last. I factored the dropped-opener detection into a shared `_has_dropped_think_opener` helper now used by both the detection path and `KimiK25Parser._recover_missing_think_open`, so masking and recovery can no longer drift.

**Review by yubofredwang — APPROVED** [review]
